subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c

```diff
@@ -0,0 +1,558 @@
+
+#define MCUMGR_IMAGE_LIST_SIZE 2
+
+struct upload_progress {
+	size_t image_size;
+	size_t offset;
+	size_t request_length;
+	size_t writed_length;
+	int image_num;
+	int upload_status;
+};
+
+struct image_read_response {
+	int slot_num;
+	int img_num;
+	struct zcbor_string hash;
+	struct zcbor_string version;
+	bool bootable;
+	bool pending;
+	bool confirmed;
+	bool active;
+	bool permanent;
+};
+
+static struct upload_progress upload_state;
+static mgmt_client_gr_image_fn client_gr_image_fn;
+static struct mcumgr_image_data image_info[MCUMGR_IMAGE_LIST_SIZE];
+static int image_count;
+static K_SEM_DEFINE(spm_mutex, 0, 1);
+static K_MUTEX_DEFINE(mcumgr_mutex);
+
+static const char smp_images_str[] = "images";
+
+static void reset_image_map_decoder_list(struct zcbor_map_decode_key_val *map, size_t map_size,
+					 struct image_read_response *buf)
+{
+	for (int i = 0; i < map_size; i++) {
+		/* Clear found status */
+		map[i].found = false;
+	}
+
+	/* Init buffer values */
+	buf->active = false;
+	buf->bootable = false;
+	buf->confirmed = false;
+	buf->pending = false;
+	buf->permanent = false;
+	buf->img_num = 0;
+	buf->slot_num = -1;
+	buf->hash.len = 0;
+	buf->version.len = 0;
+}
+
+static int image_state_res_fn(struct net_buf *nb, void *user_data)
+{
+	struct mcumgr_client_image_gr_rsp response = {0};
+	zcbor_state_t zsd[CBOR_ENCODER_STATE_NUM + 2];
+	struct zcbor_string value = {0};
+	struct image_read_response img_data;
+	struct zcbor_map_decode_key_val list_res_decode[] = {
+		/** Mandatory components */
+		ZCBOR_MAP_DECODE_KEY_DECODER("version", zcbor_tstr_decode, &img_data.version),
+		ZCBOR_MAP_DECODE_KEY_DECODER("hash", zcbor_bstr_decode, &img_data.hash),
+		ZCBOR_MAP_DECODE_KEY_DECODER("slot", zcbor_int32_decode, &img_data.slot_num),
+		/** End of mandatory components */
+		ZCBOR_MAP_DECODE_KEY_DECODER("image", zcbor_int32_decode, &img_data.img_num),
+		ZCBOR_MAP_DECODE_KEY_DECODER("bootable", zcbor_bool_decode, &img_data.bootable),
+		ZCBOR_MAP_DECODE_KEY_DECODER("pending", zcbor_bool_decode, &img_data.pending),
+		ZCBOR_MAP_DECODE_KEY_DECODER("confirmed", zcbor_bool_decode, &img_data.confirmed),
+		ZCBOR_MAP_DECODE_KEY_DECODER("active", zcbor_bool_decode, &img_data.active),
+		ZCBOR_MAP_DECODE_KEY_DECODER("permanent", zcbor_bool_decode, &img_data.permanent)};
+	bool ok;
+	int rc;
+	size_t decoded;
+
+	response.cmd = MGMT_IMAGE_STATE_RSP;
+	response.image_state.image_list_length = 0;
+	response.image_state.image_list = NULL;
+
+	if (!nb) {
+		response.status = MGMT_ERR_ETIMEOUT;
+		rc = 0;
+		goto out;
+	}
+
+	zcbor_new_decode_state(zsd, ARRAY_SIZE(zsd), nb->data, nb->len, 1);
+
+	ok = zcbor_map_start_decode(zsd);
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+
+	ok = zcbor_tstr_decode(zsd, &value);
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+	if (value.len != 6 || memcmp(value.value, smp_images_str, 6)) {
```

Comment:
What does 6 refer to?

Reply:
Ah good point It is string lenght without \0. I will do define for that.